### PR TITLE
[Testing Framework] Add test structure to views package for rendering test output

### PR DESCRIPTION
**Pull request by liamcervante**
This PR introduces a views/test.go file which adds a human-readable view for test outputs. A JSON view will be added in a later PR.

In addition, representations of test suites, test files, and test runs are added. These will be provided by the testing framework into the view to render.

## Target Release

1.6.0

## Terraform Testing Framework

This PR is part of a chain of PRs implementing the new Terraform test command.

|               Description              | Branch                     | PR |
|:--------------------------------------|----------------------------|----|
| Cleanup previous implementation        | liamcervante/tests/cleanup |  #33323  |
| Add initial human view implementation  | liamcervante/tests/view    |  #33324 |
| Add Terraform test file config support | liamcervante/tests/config  |  #33325  |
| Update terraform context functionality | liamcervante/tests/context |  #33326  |
| Add test command to Terraform CLI      | liamcervante/tests/command | #33327    |

The above PRs implement the core functionality of the new test command. More PRs will follow implementing specific features, but the test command will work end-to-end once the above PRs are merged.

These changes are implemented in a chain, with each change building on the previous one to ensure ease of review. To see the combined changes as one, navigate to the last PR in the chain and change it to merge directly into main.

### CODE REVIEW

File: internal/command/views/test.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+	counts := make(map[moduletest.Status]int)
+	for _, file := range suite.Files {
+		for _, run := range file.Runs {
+			count := counts[run.Status]
+			counts[run.Status] = count + 1
+		}
+	}
+
+	if suite.Status <= moduletest.Skip {
```
Comment (alisdair) [inline internal/command/views/test.go:65]: It feels a little odd to use inequalities against state constants. I normally don't think of iota-derived constants as having an order. Did you consider using `suite.Status == moduletest.Pending || suite.Status == moduletest.Skip` and reject it?

File: internal/moduletest/status.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+func (status Status) Merge(next Status) Status {
+	if next > status {
+		return next
+	}
+	return status
+}
+
+// ColorizedText returns textual representations of the status, but colored for
+// visual appeal.
+func (status Status) ColorizedText(color *colorstring.Colorize) string {
```
Comment (alisdair) [inline internal/moduletest/status.go:37]: This seems like view logic which I'd expect to be in the corresponding view package. Would it cause problems to move it over? If you're concerned about forgetting to add future status values we could add the views package to the `scripts/exhaustive.sh` list.

**Review by alisdair — APPROVED** [review]
This seems good to me! Some very minor thoughts inline.

File: internal/moduletest/status.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+// This should be used to collate the overall status of a test file or test
+// suite from the collection of test runs that have been executed.
+//
+// Essentially, if a test suite has a bunch of failures and passes the overall
+// status would be failure. If a test suite has all passes, then the test suite
+// would be pass overall.
+//
+// The implementation basically always returns the highest of the two, which
+// means the order the statuses are defined within the iota matters.
+func (status Status) Merge(next Status) Status {
```
Comment (liamcervante) [inline internal/moduletest/status.go:28]: @alisdair, this function does make an assumption on the ordering, basically to simplify the logic of combining multiple status. For example a test that has one error case, one fail case and two pass cases would report overall as error. So the ordering is deliberate in order to make this function simple.

File: internal/command/views/test.go (hunk 1 above)
Comment (liamcervante) [reply internal/command/views/test.go:65]: I can definitely update it to that here, but there's another function that also relies on the ordering. I added a comment there to highlight it.

Do you think it's worth me dropping any ordering assumptions to the status and updating the other function as well? I do think the iota implementation does enforce this, but I don't feel strongly about it.

Comment (alisdair) [reply internal/command/views/test.go:65]: Nope, I hadn't spotted that and I think it's fine to rely on the ordering here. I'd maybe suggest adding commentary next to the type definition to clarify that order is significant and point to `Merge` as an example of that.

**Pull request merged**